pkg/controller/velerobackup/hiveobject.go

```diff
@@ -0,0 +1,113 @@
+	switch t := object.(type) {
+	case *hivev1.ClusterDeployment:
+		checksum := calculateChecksum(t.ObjectMeta, t.Spec, controllerutils.GetChecksumOfObjects, logger)
+		return &hiveObject{
+			checksum: checksum,
+			object:   object,
+			logger:   logger,
+		}, nil
+	case *hivev1.SyncSet:
+		checksum := calculateChecksum(t.ObjectMeta, t.Spec, controllerutils.GetChecksumOfObjects, logger)
+		return &hiveObject{
+			checksum: checksum,
+			object:   object,
+			logger:   logger,
+		}, nil
+	case nil:
+		return nil, fmt.Errorf("Can't convert nil to a HiveObject")
+	}
+
+	return nil, fmt.Errorf("Unknown Type: %T", object)
+}
+
+func (h *hiveObject) hasChanged() bool {
+	if h.checksum == errChecksum {
+		// The checksum function errored. In this case we assume the object has NOT changed.
+		return false
+	}
+
+	// Make sure newChecksum is different than the old checksum.
+	oldChecksum := h.getChecksumAnnotation()
+	return h.checksum != oldChecksum
+}
+
+func calculateChecksum(meta metav1.ObjectMeta, spec interface{}, checksumFunc controllerutils.ChecksumOfObjectsFunc, logger log.FieldLogger) string {
+	// We need to take the checksum WITHOUT the previous checksum
+	// to avoid an infinite loop of the checksum always changing.
+	metaCopy := meta.DeepCopy()
+	delete(metaCopy.Annotations, controllerutils.LastBackupAnnotation)
+
+	// We also need to take the checksum WITHOUT the previous ResourceVersion
+	// as when we update the checksum annotation, ResourceVersion changes.
+	metaCopy.ResourceVersion = ""
+
+	checksum, err := checksumFunc(metaCopy, spec)
+	if err != nil {
+		// We never return an error from this function as that could lead to a
+		// situation where other objects change, but because one of the object
+		// checksum errored, then the other objects won't get backed up.
+		// Instead we return a constant checksum value. This function should rarely,
+		// if ever, return an error.
+		logger.WithError(err).Info("error calculating object checksum")
+		return errChecksum
+	}
+
+	return checksum
+}
+
+func (h *hiveObject) getChecksumAnnotation() string {
+	tmpmeta, err := meta.Accessor(h.object)
+	if err != nil {
+		// This should never happen because we should always be an object.
+		panic(err)
+	}
+	annotations := tmpmeta.GetAnnotations()
+	return annotations[controllerutils.LastBackupAnnotation]
+}
+
+func (h *hiveObject) setChecksumAnnotation(checksum string) {
+	// Not checking err because we are guaranteed to be an object.
+	tmpmeta, _ := meta.Accessor(h.object)
+	annotations := tmpmeta.GetAnnotations()
+
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// Actually set the checksum
+	annotations[controllerutils.LastBackupAnnotation] = checksum
+	tmpmeta.SetAnnotations(annotations)
+}
+
+func addHiveObjectLoggerFields(logger log.FieldLogger, hobj *hiveObject) *log.Entry {
+	return logger.WithFields(log.Fields{
+		// "name":      hobj.meta.Name,
```

Comment:
If you store a pointer to the meta in the hiveObject, then you can get the name and namespace here.

Reply:
In the last review, didn't we get rid of the pointer to the meta object in hiveObject?

I'm fine adding it back, just want to know for the future when I should store things and when I should compute things.

Reply:
Yeah, you are right. I did recommend getting rid of it. I think you'll find that I am not always consistent. Let's leave it using `meta.Accessor`. That is the better approach. Or play around with storing the runtime object as a meta.Object instead of as a runtime.Object. That would obviate the need for meta.Accessor.

Reply:
If we were to only store `meta.Object`, I don't think I can hand that to `client.Update` ?

Or is there a way to get the `runtime.object` back out of `meta.Object` (this doesn't seem likely to me as it'd be a circular reference)?

Reply:
You can cast from the `meta.Object` to the `runtime.Object`.
```
func TestObjectMeta(t *testing.T) {
	obj := &hivev1.ClusterDeployment{}
	objAsMeta, err := meta.Accessor(obj)
	if !assert.NoError(t, err, "could not convert to meta.Object") {
		return
	}
	_, ok := objAsMeta.(runtime.Object)
	assert.True(t, ok, "could not case to runtime.Object")
}
```

Reply:
Also, for what it is worth, we can probably forego using `meta.Accessor` and just cast from `runtime.Object` to `meta.Object`.

Reply:
That is super interesting. Got rid of `meta.Accessor`. Let me know if I implemented it correctly.